aplus/celery.py

```diff
@@ -32,3 +40,25 @@ def enroll():
         i.enroll_from_sis()
         if settings.SIS_ENROLL_DELAY:
             sleep(settings.SIS_ENROLL_DELAY)
+
+@app.task
+def retry_submissions():
+    from exercise.submission_models import PendingSubmission, PendingSubmissionManager
+
+    # Recovery state: only send one grading request to probe the state of grader
+    # We pick the one with most attempts, so that total_retries comes down more quickly
+    # when things are back to normal, and we can return to normal state
+    if not PendingSubmissionManager.is_grader_stable():
+        pending = PendingSubmission.objects.order_by('-num_retries').first()
+        logging.info(f"Recovery state: retrying expired submission {pending.submission}")
+        pending.submission.exercise.grade(pending.submission)
+        return
+
+    # Stable state: retry all expired submissions
+    expiry_time = datetime.datetime.now(datetime.timezone.utc) - relativedelta(seconds=settings.SUBMISSION_EXPIRY_TIMEOUT)
```

Comment:
I started to think about the default `SUBMISSION_EXPIRY_TIME` in settings.py. Should it be at least 10 minutes or even higher, 30 minutes? There are courses that use much heavier graders. Their execution times are always more than five minutes. There has been a system-wide timeout for the maximum duration. It might be around 30 minutes. (It is configured directly in Kubernetes outside A+.)